lib/vsc/mympirun/pmi/slurm.py

```diff
@@ -0,0 +1,259 @@
+from vsc.mympirun.pmi.pmi import PMI, PMIX
+from vsc.utils.run import async_to_stdout
+from vsc.mympirun.pmi.option import DISTRIBUTE_PACK, DISTRIBUTE_CYCLE
+
+DISTRIBUTE_MAP = {
+    DISTRIBUTE_PACK: 'block:block:block',
+    DISTRIBUTE_CYCLE: 'cycle:cycle:cycle',
+}
+
+
+class Slurm(Sched):
+
+    _sched_for = ['slurm']
+    SCHED_ENVIRON_ID = 'SLURM_JOB_ID'
+
+    LAUNCHER = 'srun'
+
+    def run_function(self):
+        """
+        srun itself supports redirection to file
+        """
+        args = []
+        if self.options.output:
+            args = ['--output=%s' % self.options.output]
+
+        return async_to_stdout, args
+
+    def pmicmd_sched(self):
+        """Generate the sched related arguments to the launcher as a list"""
+        args = []
+
+        # run from current directory
+        args.append("--chdir=" + os.getcwd())
+
+        if self.options.distribute:
+            args.append("--distribution=%s" % DISTRIBUTE_MAP[self.options.distribute])
+
+        return args
+
+    def pmicmd_environment(self):
+        """export everything"""
+        return ['--export=ALL']
+
+    def pmicmd_mpi(self):
+        """Handle PMI to --mpi"""
+        for pmi in self.PMI:
+            flavour = None
+            if pmi.FLAVOUR == PMIX:
+                flavour = 'pmix'
+                if pmi.VERSION == 3:
+                    flavour += '_v3'
+                else:
+                    self.log.warn("Unsupported PMIx version %s, trying with generic %s", pmi, flavour)
+            elif pmi.FLAVOUR == PMI:
+                if pmi.VERSION == 2:
+                    flavour = 'pmi2'
+                else:
+                    self.log.raiseException("Unsupported PMI version %s" % pmi)
+            else:
+                self.log.raiseException("Unsupported PMI %s" % pmi)
+
+            if flavour is not None:
+                # first one wins
+                self.log.debug("Mapped PMI %s to flavour %s", pmi, flavour)
+                return ["--mpi=%s" % flavour]
+
+        self.log.raiseException("No supported PMIs")
+        return []
+
+    def job_info(self, job_info):
+        """
+        Fill in/complete/edit job_info dict and return it
+
+        # relevant variables to use
+        SLURM_CPUS_ON_NODE=32
+        SLURM_JOB_CPUS_PER_NODE=32(x2)
+        SLURM_JOB_GPUS=0,1,2,3
+        SLURM_JOB_NODELIST=node[3302-3303]
+        SLURM_JOB_NUM_NODES=2
+        SLURM_MEM_PER_CPU=7600
+        SLURM_NNODES=2
+        SLURM_NPROCS=64
+        SLURM_NTASKS=64
+        """
+        slurm_env = {}  # slurm enviroment
+
+        dbg = []
+        prefix = 'SLURM_'
+        for key, value in os.environ.items():
+            if not key.startswith(prefix):
+                continue
+            dbg.append("%s=%s" % (key, value))
+            # some basic type conversion
+            try:
+                value = int(value)
+            except Exception as e:
+                pass
+            slurm_env[key[len(prefix):].lower()] = value
```

Comment:
You were going to clean up this `.lower()` crap, but it's still there?

Reply:
yes, i'll cleanup this mess